feature_engine/timeseries/forecasting/expanding_window_features.py

```diff
@@ -0,0 +1,232 @@
+    in the training set.
+
+    More details in the :ref:`User Guide <window_features>`.
+
+    Parameters
+    ----------
+    {variables}
+
+    min_periods: int, default None.
+        Minimum number of observations in window required to have a value;
+        otherwise, result is np.nan. See parameter `min_periods` in the pandas
+        `expanding` documentation for more details.
+
+    functions: str, list of str, default = 'mean'
+        The functions to apply within the window. Valid functions can be found
+        `here <https://pandas.pydata.org/docs/reference/window.html>`_.
+
+    sort_index: bool, default=True
+        Whether to order the index of the dataframe before creating window
+        feature.
+
+    {missing_values}
+
+    {drop_original}
+
+    Attributes
+    ----------
+    variables_:
+        The group of variables that will be used to create the window features.
+
+    {feature_names_in_}
+
+    {n_features_in_}
+
+    Methods
+    -------
+    {fit}
+
+    transform:
+        Add window features.
+
+    {fit_transform}
+
+    See Also
+    --------
+    pandas.expanding
+    pandas.aggregate
+    pandas.shift
+    """
+
+    def __init__(
+        self,
+        variables: None | int | str | list[str | int] = None,
+        min_periods: int | None = None,
+        functions: str | list[str] = "mean",
+        sort_index: bool = True,
+        missing_values: str = "raise",
+        drop_original: bool = False,
+    ) -> None:
+        if isinstance(functions, str):
+            functions = [functions]
+        if not isinstance(functions, list) or not all(
+            isinstance(val, str) for val in functions
+        ):
+            raise ValueError(
+                f"functions must be a list of strings or a string."
+                f"Got {functions} instead."
+            )
+        if len(functions) != len(set(functions)):
+            raise ValueError(
+                f"There are some duplicated functions in the list of "
+                f"functions: {functions}"
+            )
+
+        super().__init__(variables, missing_values, drop_original)
+
+        self.min_periods = min_periods
+        self.functions = functions
+        self.sort_index = sort_index
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Adds expanding window features.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The data to transform.
+
+        Returns
+        -------
+        X_new: Pandas dataframe, shape = [n_samples, n_features + window_features]
+            The dataframe with the original plus the new variables.
+        """
+        # Common dataframe checks and setting up.
+        X = super().transform(X)
+
+        tmp = (
+            X[self.variables_]
+            .expanding(min_periods=self.min_periods)
+            .agg(self.functions)
+            .shift(periods=self._infer_period(X))
```

Comment:
X is a pd.Dataframe. _infer_periods infers the period from the dataframe that we need to shift it back by. 

If you could provide an example where having the additional complexity of specifying a different period to shift by is useful then it would make sense to give the flexibility. The advantage of not having it is that it reduces the complexity of this method and simplifies testing.

Reply:
If I've got a dataframe, say with air pollutant concentrations recorded hourly. I want to predict the next 24 hours of pollutant concentration starting now. But, in reality, I only get the data from my sensors a week later after recording the air pollutants. So if I want to predict the pollutant concentration in the next 24 hs starting now, I should only use data up to a week ago, that is, my expanding window should be up to a week before the forecasting point.

Also, would infer_periods work with both integers and datetime in the index?

Reply:
I will remove infer_periods to simplify this. 

We can discuss this use case more on Monday. My concern is that we're mixing the logic of a window feature and a lag feature to cover this use case.  If that is the use case then the user should create a lag feature of the air pollutant (i.e., the air pollutant 1 week ago) and then apply a window feature to this lag feature. This cleanly separates the logic of what a window feature is and what a lag feature is and splits the code into smaller chunks which are easier to test and maintain. I'm also thinking about the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) and [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) in this context.

Reply:
To make the above point a bit more explicit. This is the use case:

```
                     ambient_temp  target_variable
2020-05-15 12:00:00         31.31            49.18
2020-05-15 12:15:00         31.51            49.84
2020-05-15 12:30:00         32.15            52.35
2020-05-15 12:45:00         32.39            50.63
2020-05-15 13:00:00         32.62            49.61
2020-05-15 13:15:00           NaN            47.01
2020-05-15 13:30:00           NaN            46.67
2020-05-15 13:45:00           NaN            47.52
```
The solution where we allow the user to specify a shift in window features is actually doing two things: 1) creating a lag feature where the user specifies the period or freq, 2) then applying the logic of what a window feature actually is.

There are potentially three solutions here:
1) The user creates a lag feature first either through LagFeatures (which currently raises an error if the variable contains a NaN) or uses `.shift()`. Then uses WindowFeatures to create a window features.
2) Create a pipeline which uses both LagFeatures and WindowFeatures (still need to deal with how NaNs are handles in LagFeatures). 
3) Mix the logic of a lag feature and window feature in the WindowFeatures class. 

Currently I don't like 3 as an option because of my reasons above (i.e., mixing logic is bad and makes things more complex to test and therefore maintain, seems to go against some common principles). The con of option 1 or 2 is that it would require a little more work from the end user. The pro for option 3 is that by putting two different pieces of logic in the same class it can reduce the number of lines of code that need to be written by the user, but makes the class more complex.

Reply:
I find that allowing to determine the end of the window will give the class more versatility / flexibility.

I am unable to gauge how many cases there are in which you can use all data up to the forecasting point. My gut tells me that it is not uncommon to have a gap between the "available data" and the "forecasting horizon". If we expose the parameters of shift, we allow users to accommodate the gap.

I do see that we can create a lag, and then do the expanding window. But with that logic, we could do the lag and then the fixed windows as well, yet we are providing that functionality out of the box in WindowFeatures.

Happy to talk more later today.

Reply:
I have made the changes as discussed:

- Freq/periods are added along with tests
- Updated the documentation to reflect this